File: LabsOff/ArrList.cs
//динамический список
public class ArrList 
{
    private int[] buf; // массив для хранения элементов 
    private int count = 0; // кол-во элементов в списке

    public ArrList()
    {
        buf = new int[5]; // задаем нач. размер
        count = 0;
    }

    // метод для добавления элементов в список
    public void Add(int a)
    {
        if (buf.Length <= count) //проверка заполности буфера
        {
            int[] newBuf = new int[buf.Length * 2]; // создаем новый буфер увеличив его(если надо) и копируем в него элементы
            for (int i = 0; i < buf.Length; i++)
            {
                newBuf[i] = buf[i];
            }
            buf = newBuf;
        }
        //добавляем элемент и увеливаем кол-во элементов
        buf[count] = a;
        count++;
    }

    // метод для добавления элемента на определенную позицию
    public void Insert(int a, int pos)
    {
        if (pos < 0 || pos > count)
        {
            return; // позиция вне диапазона списка
        }
        if (buf.Length <= count)  //проверка заполности буфера
        {
            // создаем новый буфер увеличив его(если надо) и копируем в него элементы
            int[] newBuf = new int[buf.Length * 2];
            for (int i = 0; i < buf.Length; i++)
            {
                newBuf[i] = buf[i];
            }
            buf = newBuf;
        }
        //сдвигаем элементы вправо, чтобы освободить место
        for (int i = count; i > pos; i--)
        {
            buf[i] = buf[i - 1];
        }
        //вставляем элемент на указанную позицию и увеличиваем кол-во элементов 
        buf[pos] = a;
        count++;
    }

    // метод для удаления элемента с определенной позиции
    public void Delete(int pos)
    {
        if (pos < 0 || pos >= count)
        {
            return; //проверка позиции
        }
        //сдвигаем элементы влево для удаления
        for (int i = pos; i < count - 1; i++)
        {
            buf[i] = buf[i + 1];
        }
        //уменьшаем кол-во элементов
        count--;
    }

    // метод очистки списка
    public void Clear()
    {
        for (int i = 0; i < count; i++)
        {
            buf[i] = 0; //зануляем все элементы
        }
        count = 0; //сбрасываем кол-во элементов
    }

    //метод для вывода элементов списка (в столбик)
    public void Print()
    {
        Console.WriteLine("Динамический массив:");
        for (int i = 0; i < count; i++)
        {
            Console.Write(buf[i] + " ");
        }
        Console.WriteLine();
    }
        
    // получение кол-ва элементов в списке
    public int Count
    {
        get
        {
            return count;
        }
    }

    //индексатор для доступа к элементам списка
    public int this[int i]
    {
        get
        {
            if (i >= count || i < 0)
            {
                throw new ArgumentOutOfRangeException("Element is out of range");
            }
            return buf[i]; //возращаем элемент списка по указанному индексу
        }

        set
        {
            if (i >= count || i < 0)
            {
                //throw new ArgumentOutOfRangeException("Element is out of range");
                return;
            }
            buf[i] = value; //присваиваем новое значение элементу списка по указанному индексу
        }
    }

    public void deleteDublicates()
    {
        if (count <= 0) //проверяем на кол-во элементов в списке
        {
            return;
        }
        for (int i = 0; i < count; i++)
        {
            int currentElement = this[i];
            bool isDublicate = false;
            for (int j = count - 1; j > i; j--)
            {
                if (this[j] == currentElement)
                {
                    Delete(j);
                    isDublicate = true;
                }
            }
            if (isDublicate == true)
            {
                Delete(i);
                i--;
            }
        }
    }
}

File: lab4/laba1/ArrList.cpp
#include "ArrList.h"

ArrList::ArrList() {
    capacity = 5;
    buf = new int*[capacity]; // Выделяем память под массив указателей
    for (int i = 0; i < capacity; ++i) {
        buf[i] = new int; // Выделяем память для каждого элемента массива
    }
    count = 0;
}

void ArrList::Add(int data) {
    if (count >= capacity) {
        // Увеличиваем емкость массива указателей
        int newCapacity = capacity * 2;
        int** newBuf = new int*[newCapacity];
        for (int i = 0; i < newCapacity; ++i) {
            if (i < capacity) {
                newBuf[i] = buf[i]; // Копируем указатели из старого буфера
            } else {
                newBuf[i] = new int; // Выделяем новую память для дополнительных элементов
            }
        }
        delete[] buf; // Освобождаем память для старого массива указателей
        buf = newBuf;
        capacity = newCapacity;
    }
    *(buf[count]) = data; // Записываем значение по адресу, хранящемуся в buf[count]
    count++;
}

void ArrList::Insert(int data, int pos) {
    if (pos < 0 || pos > count) return;
    if (count >= capacity) {
        // Увеличиваем емкость массива указателей аналогично методу Add
        int newCapacity = capacity * 2;
        int** newBuf = new int*[newCapacity];
        for (int i = 0; i < newCapacity; ++i) {
            if (i < capacity) {
                newBuf[i] = buf[i];
            } else {
                newBuf[i] = new int;
            }
        }
        delete[] buf;
        buf = newBuf;
        capacity = newCapacity;
    }
    for (int i = count; i > pos; --i) {
        *(buf[i]) = *(buf[i - 1]); // Сдвигаем элементы вправо
    }
    *(buf[pos]) = data; // Вставляем новое значение
    count++;
}

void ArrList::Delete(int pos) {
    if (pos < 0 || pos >= count) return;
    for (int i = pos; i < count - 1; ++i) {
        *(buf[i]) = *(buf[i + 1]); // Сдвигаем элементы влево
    }
    count--;
}

void ArrList::Clear() {
    count = 0; // Просто обнуляем количество элементов
}

int ArrList::Count() {
    return count;
}

int& ArrList::operator[](int i) {
    if (i < 0 || i >= count) {
        throw std::out_of_range("Index out of range");
    }
    return *(buf[i]); // Возвращаем значение по адресу, хранящемуся в buf[i]
}

void ArrList::Print() {
    for (int i = 0; i < count; ++i) {
        std::cout << *(buf[i]) << " "; // Выводим значение по адресу, хранящемуся в buf[i]
    }
    std::cout << std::endl;
}